File: keymagic-windows/gui-tauri/src-tauri/src/hud.rs
use std::sync::Mutex;
use std::sync::OnceLock;
use std::ffi::c_void;
use windows::core::*;
use windows::Win32::Foundation::*;
use windows::Win32::Graphics::Gdi::*;
use windows::Win32::System::LibraryLoader::*;
use windows::Win32::UI::WindowsAndMessaging::*;
use anyhow::{Result, anyhow};
use log::{debug, error};

static HUD_INSTANCE: OnceLock<Mutex<HWND>> = OnceLock::new();

const HUD_TIMER_ID: usize = 1;
const HUD_DISPLAY_TIME_MS: u32 = 1500;
const WM_SHOW_HUD: u32 = WM_USER + 1;
const AC_SRC_OVER: u8 = 0x00;
const AC_SRC_ALPHA: u8 = 0x01;

/// Initialize the HUD window
pub fn initialize_hud() -> Result<()> {
    debug!("Initializing HUD window");
    unsafe {
        let instance = GetModuleHandleW(None)?;
        debug!("Got module handle");
        
        // Register window class
        let class_name = w!("KeyMagicHUD");
        
        let wc = WNDCLASSEXW {
            cbSize: std::mem::size_of::<WNDCLASSEXW>() as u32,
            style: CS_HREDRAW | CS_VREDRAW,
            lpfnWndProc: Some(hud_wnd_proc),
            cbClsExtra: 0,
            cbWndExtra: std::mem::size_of::<*mut c_void>() as i32,
            hInstance: instance.into(),
            hIcon: HICON::default(),
            hCursor: LoadCursorW(HINSTANCE::default(), IDC_ARROW)?,
            hbrBackground: HBRUSH((COLOR_WINDOW.0 + 1) as isize),
            lpszMenuName: PCWSTR::null(),
            lpszClassName: class_name,
            hIconSm: HICON::default(),
        };

        let atom = RegisterClassExW(&wc);
        if atom == 0 {
            let err = GetLastError();
            if err != Ok(()) {
                // Check if it's not already registered
                // For now, continue even if registration failed
            }
        }

        // Create window
        let hwnd = CreateWindowExW(
            WS_EX_LAYERED | WS_EX_TRANSPARENT | WS_EX_TOPMOST | WS_EX_TOOLWINDOW | WS_EX_NOACTIVATE,
            class_name,
            w!(""),
            WS_POPUP,
            0, 0, 0, 0,
            HWND::default(),
            HMENU::default(),
            instance,
            None,
        );

        if hwnd.0 == 0 {
            return Err(anyhow!("Failed to create HUD window"));
        }
        
        debug!("HUD window created successfully: {:?}", hwnd);

        // Store window handle
        HUD_INSTANCE.set(Mutex::new(hwnd))
            .map_err(|_| anyhow!("Failed to store HUD window handle"))?;
            
        debug!("HUD initialization complete");

        Ok(())
    }
}

/// Show the HUD with keyboard name
pub fn show_keyboard_hud(keyboard_name: &str) -> Result<()> {
    debug!("show_keyboard_hud called with: {}", keyboard_name);
    if let Some(hwnd_mutex) = HUD_INSTANCE.get() {
        if let Ok(hwnd) = hwnd_mutex.lock() {
            unsafe {
                debug!("Sending HUD message for: {}", keyboard_name);
                // Allocate and copy string
                let text_wide: Vec<u16> = keyboard_name.encode_utf16().chain(std::iter::once(0)).collect();
                let text_ptr = Box::into_raw(Box::new(text_wide));
                
                // Send message to show HUD
                let result = PostMessageW(*hwnd, WM_SHOW_HUD, WPARAM(0), LPARAM(text_ptr as isize));
                if result.is_err() {
                    error!("Failed to post message: {:?}", result);
                }
            }
        } else {
            error!("Failed to lock HUD mutex");
        }
    } else {
        error!("HUD not initialized");
    }
    Ok(())
}


/// Show tray minimize notification
pub fn show_tray_minimize_notification() -> Result<()> {
    show_keyboard_hud("KeyMagic is running in the system tray")
}

/// Window procedure for HUD
extern "system" fn hud_wnd_proc(hwnd: HWND, msg: u32, wparam: WPARAM, lparam: LPARAM) -> LRESULT {
    unsafe {
        match msg {
            WM_SHOW_HUD => {
                debug!("WM_SHOW_HUD received");
                // Get keyboard name from lparam
                let text_ptr = lparam.0 as *mut Vec<u16>;
                if !text_ptr.is_null() {
                    let text_vec = Box::from_raw(text_ptr);
                    let text_str = String::from_utf16_lossy(&text_vec);
                    debug!("Showing HUD for: {}", text_str);
                    
                    // Show HUD
                    show_hud_internal(hwnd, &text_vec);
                    
                    // Set timer
                    let _ = SetTimer(hwnd, HUD_TIMER_ID, HUD_DISPLAY_TIME_MS, None);
                }
                LRESULT(0)
            }
            WM_TIMER => {
                if wparam.0 == HUD_TIMER_ID {
                    hide_hud(hwnd);
                    let _ = KillTimer(hwnd, HUD_TIMER_ID);
                }
                LRESULT(0)
            }
            WM_NCHITTEST => {
                // Make window click-through
                LRESULT(-1)  // HTNOWHERE
            }
            _ => DefWindowProcW(hwnd, msg, wparam, lparam),
        }
    }
}

/// Internal function to show HUD
fn show_hud_internal(hwnd: HWND, text: &[u16]) {
    debug!("show_hud_internal called");
    unsafe {
        // First make the window visible
        let _ = ShowWindow(hwnd, SW_SHOWNOACTIVATE);

        let hdc = GetDC(HWND_DESKTOP);
        let mem_dc = CreateCompatibleDC(hdc);
        
        // Create font
        let font = CreateFontW(
            -mul_div(20, GetDeviceCaps(hdc, LOGPIXELSY), 72),
            0, 0, 0,
            FW_NORMAL.0 as i32,
            FALSE.0 as u32,
            FALSE.0 as u32,
            FALSE.0 as u32,
            DEFAULT_CHARSET.0 as u32,
            OUT_DEFAULT_PRECIS.0 as u32,
            CLIP_DEFAULT_PRECIS.0 as u32,
            DEFAULT_QUALITY.0 as u32,
            DEFAULT_PITCH.0 as u32 | FF_SWISS.0 as u32,
            w!("Segoe UI"),
        );
        
        let _old_font = SelectObject(mem_dc, font);
        
        // Measure text
        let mut size = SIZE::default();
        let _ = GetTextExtentPoint32W(mem_dc, text, &mut size);
        
        let padding = 20;
        let width = size.cx + (padding * 2);
        let height = size.cy + (padding * 2);
        
        // Create bitmap
        let bitmap = CreateCompatibleBitmap(hdc, width, height);
        let _old_bitmap = SelectObject(mem_dc, bitmap);
        
        // Define colors
        let transparent_color = rgb(255, 0, 255); // Magenta for transparency
        let bg_color = rgb(0, 0, 0); // Black background
        let text_color = rgb(255, 255, 255); // White text
        
        // First fill entire bitmap with transparent color
        let transparent_pen = CreatePen(PS_SOLID, 0, transparent_color);
        let _ = SelectObject(mem_dc, transparent_pen);
        let transparent_brush = CreateSolidBrush(transparent_color);
        let _ = SelectObject(mem_dc, transparent_brush);
        let _ = Rectangle(mem_dc, 0, 0, width, height);
        
        // Draw rounded rectangle with black background
        let bg_brush = CreateSolidBrush(bg_color);
        let _ = SelectObject(mem_dc, bg_brush);
        let _ = RoundRect(mem_dc, 0, 0, width, height, 22, 22);
        
        // Draw text
        let _ = SetBkMode(mem_dc, TRANSPARENT);
        let _ = SetTextColor(mem_dc, text_color);
        let rect = RECT { left: 0, top: 0, right: width, bottom: height };
        let mut text_rect = rect.clone();
        let mut text_copy: Vec<u16> = text.to_vec();
        let _ = DrawTextW(mem_dc, &mut text_copy[..], &mut text_rect as *mut _, DT_CENTER | DT_SINGLELINE | DT_VCENTER);
        
        // Apply alpha channel similar to C++ sample
        set_bitmap_alpha(mem_dc, bitmap, transparent_color, text_color);
        
        // Cleanup
        let _ = DeleteObject(transparent_pen);
        let _ = DeleteObject(transparent_brush);
        let _ = DeleteObject(bg_brush);
        
        // Update layered window
        update_layered_window(hwnd, mem_dc, width, height);
        
        // Cleanup
        let _ = DeleteObject(font);
        let _ = DeleteObject(bitmap);
        let _ = DeleteDC(mem_dc);
        let _ = ReleaseDC(HWND_DESKTOP, hdc);
    }
}

/// Update layered window with bitmap
fn update_layered_window(hwnd: HWND, mem_dc: HDC, width: i32, height: i32) {
    debug!("update_layered_window called with size: {}x{}", width, height);
    unsafe {
        // Get monitor info
        let monitor = MonitorFromWindow(hwnd, MONITOR_DEFAULTTOPRIMARY);
        let mut monitor_info = MONITORINFO {
            cbSize: std::mem::size_of::<MONITORINFO>() as u32,
            ..Default::default()
        };
        let _ = GetMonitorInfoW(monitor, &mut monitor_info);
        
        // Calculate position - bottom-right corner
        let work_area = monitor_info.rcWork;
        let margin = 20;
        let x = work_area.right - width - margin;
        let y = work_area.bottom - height - margin - 50;
        
        let size = SIZE { cx: width, cy: height };
        let src_point = POINT { x: 0, y: 0 };
        let dst_point = POINT { x, y };
        
        let blend = BLENDFUNCTION {
            BlendOp: AC_SRC_OVER as u8,
            BlendFlags: 0,
            SourceConstantAlpha: 255,  // Use per-pixel alpha
            AlphaFormat: AC_SRC_ALPHA as u8,  // Important: use per-pixel alpha
        };
        
        let transparent_color = rgb(255, 0, 255); // Magenta for transparency
        
        let _ = UpdateLayeredWindow(
            hwnd,
            HDC::default(),
            Some(&dst_point),
            Some(&size),
            mem_dc,
            Some(&src_point),
            transparent_color,  // Use transparent color like C++ sample
            Some(&blend),
            ULW_ALPHA,  // Use ULW_ALPHA like C++ sample
        );
    }
}

/// Hide the HUD
fn hide_hud(hwnd: HWND) {
    unsafe {
        // Hide by moving off-screen
        let _ = SetWindowPos(
            hwnd,
            HWND_TOP,
            -1000, -1000, 0, 0,
            SWP_HIDEWINDOW | SWP_NOACTIVATE,
        );
    }
}

// MulDiv implementation for Windows
fn mul_div(n_number: i32, n_numerator: i32, n_denominator: i32) -> i32 {
    ((n_number as i64 * n_numerator as i64) / n_denominator as i64) as i32
}

// RGB macro to create COLORREF
fn rgb(r: u8, g: u8, b: u8) -> COLORREF {
    COLORREF((r as u32) | ((g as u32) << 8) | ((b as u32) << 16))
}

// Set bitmap alpha based on colors (matching C++ sample)
fn set_bitmap_alpha(hdc: HDC, bitmap: HBITMAP, transparent_color: COLORREF, text_color: COLORREF) {
    unsafe {
        let mut bm = BITMAP::default();
        GetObjectW(bitmap, std::mem::size_of::<BITMAP>() as i32, Some(&mut bm as *mut _ as *mut c_void));
        
        // Create BITMAPINFO
        let bmi_size = std::mem::size_of::<BITMAPINFOHEADER>() + (256 * std::mem::size_of::<RGBQUAD>());
        let mut bmi_buffer = vec![0u8; bmi_size];
        let bmi = bmi_buffer.as_mut_ptr() as *mut BITMAPINFO;
        (*bmi).bmiHeader.biSize = std::mem::size_of::<BITMAPINFOHEADER>() as u32;
        
        // Get bitmap info
        let _ = GetDIBits(hdc, bitmap, 0, bm.bmHeight as u32, None, bmi, DIB_RGB_COLORS);
        
        if (*bmi).bmiHeader.biBitCount != 32 {
            return;
        }
        
        // Allocate buffer for pixel data
        let pixel_count = (bm.bmWidth * bm.bmHeight) as usize;
        let mut pixel_data = vec![0u8; pixel_count * 4];
        
        // Get pixel data
        let _ = GetDIBits(hdc, bitmap, 0, bm.bmHeight as u32, Some(pixel_data.as_mut_ptr() as *mut c_void), bmi, DIB_RGB_COLORS);
        
        // Process each pixel
        for i in 0..pixel_count {
            let offset = i * 4;
            let b = pixel_data[offset];
            let g = pixel_data[offset + 1];
            let r = pixel_data[offset + 2];
            let color = rgb(r, g, b);
            
            let alpha = if color == transparent_color {
                0u8
            } else if color == text_color {
                255u8
            } else {
                (255.0 * 0.8) as u8 // 80% opacity for background
            };
            
            pixel_data[offset + 3] = alpha;
            // Premultiply alpha
            pixel_data[offset] = (b as u32 * alpha as u32 / 255) as u8;
            pixel_data[offset + 1] = (g as u32 * alpha as u32 / 255) as u8;
            pixel_data[offset + 2] = (r as u32 * alpha as u32 / 255) as u8;
        }
        
        // Set the modified pixels back
        let _ = SetDIBits(hdc, bitmap, 0, bm.bmHeight as u32, pixel_data.as_ptr() as *const c_void, bmi, DIB_RGB_COLORS);
    }
}

File: keymagic-windows/tsf/src/HUD.cpp
#include "HUD.h"
#include <algorithm>
#include <vector>

// RGB macro
#define RGB_MACRO(r,g,b) ((COLORREF)(((BYTE)(r)|((WORD)((BYTE)(g))<<8))|(((DWORD)(BYTE)(b))<<16)))

KeyMagicHUD& KeyMagicHUD::GetInstance()
{
    static KeyMagicHUD instance;
    return instance;
}

KeyMagicHUD::KeyMagicHUD() : m_hwnd(nullptr)
{
}

KeyMagicHUD::~KeyMagicHUD()
{
    Cleanup();
}

HRESULT KeyMagicHUD::Initialize()
{
    if (m_hwnd != nullptr)
        return S_OK; // Already initialized
        
    HINSTANCE hInstance = GetModuleHandle(nullptr);
    
    // Register window class
    WNDCLASSEXW wc = {};
    wc.cbSize = sizeof(WNDCLASSEXW);
    wc.style = CS_HREDRAW | CS_VREDRAW;
    wc.lpfnWndProc = WndProc;
    wc.cbWndExtra = sizeof(void*);
    wc.hInstance = hInstance;
    wc.hCursor = LoadCursor(nullptr, IDC_ARROW);
    wc.hbrBackground = (HBRUSH)(COLOR_WINDOW + 1);
    wc.lpszClassName = L"KeyMagicHUD";
    
    ATOM atom = RegisterClassExW(&wc);
    if (atom == 0 && GetLastError() != ERROR_CLASS_ALREADY_EXISTS)
    {
        return HRESULT_FROM_WIN32(GetLastError());
    }
    
    // Create window
    m_hwnd = CreateWindowExW(
        WS_EX_LAYERED | WS_EX_TRANSPARENT | WS_EX_TOPMOST | WS_EX_TOOLWINDOW | WS_EX_NOACTIVATE,
        L"KeyMagicHUD",
        L"",
        WS_POPUP,
        0, 0, 0, 0,
        nullptr,
        nullptr,
        hInstance,
        this // Pass this pointer for WM_CREATE
    );
    
    if (!m_hwnd)
    {
        return HRESULT_FROM_WIN32(GetLastError());
    }
    
    return S_OK;
}

void KeyMagicHUD::ShowKeyboard(const std::wstring& keyboardName)
{
    if (!m_hwnd)
        return;
        
    // Allocate and copy string
    std::wstring* pText = new std::wstring(keyboardName);
    
    // Post message to show HUD
    PostMessage(m_hwnd, WM_SHOW_HUD, 0, reinterpret_cast<LPARAM>(pText));
}

void KeyMagicHUD::Cleanup()
{
    if (m_hwnd)
    {
        DestroyWindow(m_hwnd);
        m_hwnd = nullptr;
    }
}

LRESULT CALLBACK KeyMagicHUD::WndProc(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam)
{
    KeyMagicHUD* pThis = nullptr;
    
    if (msg == WM_CREATE)
    {
        CREATESTRUCT* pCreate = reinterpret_cast<CREATESTRUCT*>(lParam);
        pThis = reinterpret_cast<KeyMagicHUD*>(pCreate->lpCreateParams);
        SetWindowLongPtr(hwnd, GWLP_USERDATA, reinterpret_cast<LONG_PTR>(pThis));
    }
    else
    {
        pThis = reinterpret_cast<KeyMagicHUD*>(GetWindowLongPtr(hwnd, GWLP_USERDATA));
    }
    
    if (pThis)
    {
        return pThis->HandleMessage(hwnd, msg, wParam, lParam);
    }
    
    return DefWindowProcW(hwnd, msg, wParam, lParam);
}

LRESULT KeyMagicHUD::HandleMessage(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam)
{
    switch (msg)
    {
        case WM_SHOW_HUD:
        {
            std::wstring* pText = reinterpret_cast<std::wstring*>(lParam);
            if (pText)
            {
                ShowHudInternal(*pText);
                delete pText;
                
                // Set timer to hide
                SetTimer(hwnd, HUD_TIMER_ID, HUD_DISPLAY_TIME_MS, nullptr);
            }
            return 0;
        }
        
        case WM_TIMER:
        {
            if (wParam == HUD_TIMER_ID)
            {
                HideHud();
                KillTimer(hwnd, HUD_TIMER_ID);
            }
            return 0;
        }
        
        case WM_NCHITTEST:
            return HTNOWHERE; // Make window click-through
            
        default:
            return DefWindowProcW(hwnd, msg, wParam, lParam);
    }
}

void KeyMagicHUD::ShowHudInternal(const std::wstring& text)
{
    // Show window
    ShowWindow(m_hwnd, SW_SHOWNOACTIVATE);
    
    HDC hdcScreen = GetDC(nullptr);
    HDC memDC = CreateCompatibleDC(hdcScreen);
    
    // Create font
    int fontSize = -MulDiv(20, GetDeviceCaps(hdcScreen, LOGPIXELSY), 72);
    HFONT font = CreateFontW(
        fontSize, 0, 0, 0,
        FW_NORMAL, FALSE, FALSE, FALSE,
        DEFAULT_CHARSET, OUT_DEFAULT_PRECIS,
        CLIP_DEFAULT_PRECIS, DEFAULT_QUALITY,
        DEFAULT_PITCH | FF_SWISS,
        L"Segoe UI"
    );
    
    HFONT oldFont = (HFONT)SelectObject(memDC, font);
    
    // Measure text
    SIZE textSize;
    GetTextExtentPoint32W(memDC, text.c_str(), (int)text.length(), &textSize);
    
    const int padding = 20;
    int width = textSize.cx + (padding * 2);
    int height = textSize.cy + (padding * 2);
    
    // Create bitmap
    HBITMAP bitmap = CreateCompatibleBitmap(hdcScreen, width, height);
    HBITMAP oldBitmap = (HBITMAP)SelectObject(memDC, bitmap);
    
    // Colors
    COLORREF transparentColor = RGB_MACRO(255, 0, 255); // Magenta
    COLORREF bgColor = RGB_MACRO(0, 0, 0); // Black
    COLORREF textColor = RGB_MACRO(255, 255, 255); // White
    
    // Fill with transparent color
    HBRUSH transparentBrush = CreateSolidBrush(transparentColor);
    RECT fillRect = {0, 0, width, height};
    FillRect(memDC, &fillRect, transparentBrush);
    DeleteObject(transparentBrush);
    
    // Draw rounded rectangle background
    HBRUSH bgBrush = CreateSolidBrush(bgColor);
    HBRUSH oldBrush = (HBRUSH)SelectObject(memDC, bgBrush);
    HPEN bgPen = CreatePen(PS_SOLID, 0, bgColor);
    HPEN oldPen = (HPEN)SelectObject(memDC, bgPen);
    
    RoundRect(memDC, 0, 0, width, height, 22, 22);
    
    SelectObject(memDC, oldBrush);
    SelectObject(memDC, oldPen);
    DeleteObject(bgBrush);
    DeleteObject(bgPen);
    
    // Draw text
    SetBkMode(memDC, TRANSPARENT);
    SetTextColor(memDC, textColor);
    RECT textRect = {0, 0, width, height};
    DrawTextW(memDC, text.c_str(), (int)text.length(), &textRect, DT_CENTER | DT_SINGLELINE | DT_VCENTER);
    
    // Apply alpha channel
    SetBitmapAlpha(memDC, bitmap, transparentColor, textColor);
    
    // Update layered window
    UpdateLayeredWindow(memDC, width, height);
    
    // Cleanup
    SelectObject(memDC, oldFont);
    SelectObject(memDC, oldBitmap);
    DeleteObject(font);
    DeleteObject(bitmap);
    DeleteDC(memDC);
    ReleaseDC(nullptr, hdcScreen);
}

void KeyMagicHUD::HideHud()
{
    // Hide by moving off-screen
    SetWindowPos(m_hwnd, HWND_TOP, -1000, -1000, 0, 0, SWP_HIDEWINDOW | SWP_NOACTIVATE);
}

void KeyMagicHUD::UpdateLayeredWindow(HDC memDC, int width, int height)
{
    // Get monitor info
    HMONITOR monitor = MonitorFromWindow(m_hwnd, MONITOR_DEFAULTTOPRIMARY);
    MONITORINFO mi = {};
    mi.cbSize = sizeof(MONITORINFO);
    GetMonitorInfo(monitor, &mi);
    
    // Calculate position - bottom-right corner
    const int margin = 20;
    int x = mi.rcWork.right - width - margin;
    int y = mi.rcWork.bottom - height - margin - 50;
    
    SIZE size = {width, height};
    POINT srcPoint = {0, 0};
    POINT dstPoint = {x, y};
    
    BLENDFUNCTION blend = {};
    blend.BlendOp = AC_SRC_OVER;
    blend.BlendFlags = 0;
    blend.SourceConstantAlpha = 255;
    blend.AlphaFormat = AC_SRC_ALPHA;
    
    COLORREF transparentColor = RGB_MACRO(255, 0, 255);
    
    ::UpdateLayeredWindow(
        m_hwnd,
        nullptr,
        &dstPoint,
        &size,
        memDC,
        &srcPoint,
        transparentColor,
        &blend,
        ULW_ALPHA
    );
}

void KeyMagicHUD::SetBitmapAlpha(HDC hdc, HBITMAP bitmap, COLORREF transparentColor, COLORREF textColor)
{
    BITMAP bm;
    GetObject(bitmap, sizeof(BITMAP), &bm);
    
    BITMAPINFOHEADER bih = {};
    bih.biSize = sizeof(BITMAPINFOHEADER);
    bih.biWidth = bm.bmWidth;
    bih.biHeight = bm.bmHeight;
    bih.biPlanes = 1;
    bih.biBitCount = 32;
    bih.biCompression = BI_RGB;
    
    BITMAPINFO bi = {};
    bi.bmiHeader = bih;
    
    // Get pixel data
    int pixelCount = bm.bmWidth * bm.bmHeight;
    std::vector<BYTE> pixelData(pixelCount * 4);
    
    GetDIBits(hdc, bitmap, 0, bm.bmHeight, pixelData.data(), &bi, DIB_RGB_COLORS);
    
    // Process each pixel
    for (int i = 0; i < pixelCount; i++)
    {
        int offset = i * 4;
        BYTE b = pixelData[offset];
        BYTE g = pixelData[offset + 1];
        BYTE r = pixelData[offset + 2];
        COLORREF color = RGB_MACRO(r, g, b);
        
        BYTE alpha;
        if (color == transparentColor)
        {
            alpha = 0;
        }
        else if (color == textColor)
        {
            alpha = 255;
        }
        else
        {
            alpha = (BYTE)(255 * 0.8); // 80% opacity for background
        }
        
        pixelData[offset + 3] = alpha;
        // Premultiply alpha
        pixelData[offset] = (BYTE)(b * alpha / 255);
        pixelData[offset + 1] = (BYTE)(g * alpha / 255);
        pixelData[offset + 2] = (BYTE)(r * alpha / 255);
    }
    
    // Set the modified pixels back
    SetDIBits(hdc, bitmap, 0, bm.bmHeight, pixelData.data(), &bi, DIB_RGB_COLORS);
}